desktop/core/src/desktop/lib/importer/operations.py

```diff
@@ -0,0 +1,685 @@
+      return {
+        'type': file_type,
+        'has_header': has_header,
+        'columns': [],
+        'preview_data': {},
+      }
+
+    # Get schema and preview data
+    schema = df.schema
+
+    # Limit preview to specified number of rows for performance
+    preview_data = df.head(preview_rows).to_dict(as_series=False)
+
+    # Create column metadata with SQL type mapping
+    columns = []
+    for i, col in enumerate(df.columns):
+      col_type = str(schema[col])
+      sql_type = map_polars_to_sql(dialect, col_type)
+
+      columns.append(
+        {
+          'name': col,
+          'type': sql_type,
+          'index': i,  # Add index for ordering
+        }
+      )
+
+    result = {
+      'type': file_type,
+      'has_header': has_header,
+      'columns': columns,
+      'preview_data': preview_data,
+      'total_rows': df.height,  # Include total row count
+      'sheet_name': sheet_name,  # Include sheet name in result
+    }
+
+    return result
+
+  except Exception as e:
+    message = f"Error previewing Excel file: {e}"
+    LOG.error(message, exc_info=True)
+    raise Exception(message)
+
+
+def _preview_delimited_file(
+  fh: BinaryIO,
+  file_type: str,
+  field_separator: str,
+  quote_char: str,
+  record_separator: str,
+  dialect: str,
+  has_header: Optional[bool],
+  preview_rows: int = 50,
+) -> Dict[str, Any]:
+  """Preview a delimited file (CSV, TSV, etc.)
+
+  Args:
+    fh: File handle for the delimited file
+    file_type: Type of file ('csv', 'tsv', 'delimiter_format')
+    field_separator: Field separator character
+    quote_char: Quote character
+    record_separator: Record separator character
+    dialect: SQL dialect for type mapping
+    has_header: Whether the file has a header row (None means auto-detect)
+    preview_rows: Number of rows to include in preview (default: 50)
+
+  Returns:
+    Dict containing:
+      - type: File type
+      - has_header: Whether file has headers
+      - columns: List of column metadata
+      - preview_data: Preview of file data
+      - total_rows: Approximate total row count
+      - delimiters: Information about detected delimiters
+
+  Raises:
+    Exception: If there's an error processing the delimited file
+  """
+  try:
+    # Auto-detect header if not specified
+    if has_header is None:
+      fh.seek(0)
+      try:
+        # Read 16 KiB sample
+        sample = fh.read(16 * 1024).decode('utf-8', errors='replace')
+
+        has_header = csv.Sniffer().has_header(sample)
+        LOG.info(f"Auto-detected header for delimited file: {has_header}")
+      except Exception as e:
+        LOG.warning(f"Error auto-detecting header in delimited file: {e}")
+        # Default based on file type - most CSVs have headers
+        has_header = True if file_type in ['csv', 'tsv'] else False
+    else:
+      has_header = coerce_bool(has_header)
+
+    # Reset file position and read file content
+    fh.seek(0)
+
+    df = pl.read_csv(
```

Comment:
Does this read the whole file in memory? Or can we restrict to to certain rows/bytes?

Reply:
Nice catch! Fixed it now with new commit